core/app/models/spree/notification_dispatch.rb

```diff
@@ -0,0 +1,81 @@
+module Spree
+  # This class is responsible for sending 'transactional' emails to users,
+  # carton shipped, order confirmed, etc.
+  #
+  # By default it uses the NotificationDispatch::ActionMailerDispatch to
+  # send emails with ActionMailer. You can change the actual delivery method
+  # by changing `Spree::NotificationDispatch.delivery_class_name` to the name
+  # of a custom class, that sends email with a different API, or even sends
+  # messages by means other than email.
+  #
+  # The sender class should have the same API as the NotificationDispatch:
+  #
+  #    SenderClass.new(message_type_symbol).deliver(messsage, arg1, arg2)
+  #
+  # To add a new notification message, edit NotificationDispatch.signatures
+  # to add the signature, and if using the ActionMailerDispatch, edit
+  # mailer_dispatch_table there so it knows how to handle it.
+  #
+  # All messages will be suppressed if Spree::Config[:send_core_emails] is false.
+  # additionally you can send only certain messages with eg:
+  #
+  #     Spree::NotificationDispatch.only_messages = [:order_confirm, :carton_shipped]
+  #
+  # or
+  #
+  #     Spree::NotificationDispatch.except_messages = [:reimbursement_processed]
+  class NotificationDispatch
+    # These serve as the specififications of what arguments go with what
+    # notification messages, and are checked at runtime.
+    class_attribute :signatures
+    # the lambda bodies could be non-empty and check actual types if we want.
+    self.signatures = {
+      # should be (carton:, order:, resend: false), but currently supporting
+      # deprecated options.
+      carton_shipped: lambda { |options, deprecated_options = {}| },
+      order_confirm: lambda { |order, resend = false| },
+      order_cancel: lambda { |order, resend = false| },
+      order_inventory_cancel: lambda { |order, inventory_units, resend = false| },
+      reimbursement_processed: lambda { |reimbursement, resend = false| }
+    }
```

Comment:
Without this, if I want to write a new back-end replacing ActionMailer delivery with another delivery mechanism (likely an HTTP api, say, mailgun; this was one of the design goals suggested by jhawthorn which I agree with) -- where would I go to look at exactly what messages my new back-end needs to respond to?

Just the existing ActionMailer-based implementation?  While it's the default, I don't think it should be privileged like that, I think a specification of what the different notifications are, and what their signature sare, needs to be somewhere that is not coupled to ActionMailer. 

That is the only and entire point of the `self.signatures` stuff.  

Especially because these signatures can and have evolved in the past (both new ones added, and existing ones had their parameters changed), and will continue to do so in the future, I think a central specification of exactly what the notifications are and what their parameters are is needed to allow alternate non-ActionMailer back-ends to be written in a sustainable way, that allows the signatures to continue to evolve in a way that's maintainable for alternate non-ActionMailer back-ends that may be in local code or plugins.